File: src/Core/StagingFinancialEvents/Services/Transformers/LotProcessors/StockLotProcessor.py
from typing import Sequence

import arrow as ar

import Core.FinancialEvents.Schemas.CommonFormats as cf
import Core.FinancialEvents.Schemas.Grouping as pgf
from Core.StagingFinancialEvents.Contracts.LotProcessor import LotProcessor
from Core.StagingFinancialEvents.Schemas.Lots import StagingTaxLot


class StockLotProcessor(LotProcessor[StagingTaxLot, pgf.TaxLotStock, Sequence[pgf.TradeEventStockAcquired | pgf.TradeEventStockSold]]):

    def process(
        self,
        input: StagingTaxLot,
        references: Sequence[pgf.TradeEventStockAcquired | pgf.TradeEventStockSold],
    ) -> pgf.TaxLotStock:
        # print("Processing stock lot (ID: {})".format(lot.ID))

        allBuys: list[pgf.TradeEventStockAcquired] = list(filter(lambda trade: isinstance(trade, pgf.TradeEventStockAcquired), references))  # type: ignore
        allSells: list[pgf.TradeEventStockSold] = list(filter(lambda trade: isinstance(trade, pgf.TradeEventStockSold), references))  # type: ignore

        # TODO: Validate returns since buys and sells are merged
        # TODO: What to do when no match is found?
        try:
            matchingBuyById = self.utils.findStockEventById(input.Acquired.ID or "", allBuys)
            matchingSoldByDate = self.utils.findStockEventByDate(input.Sold.DateTime or ar.get("1-0-0"), allSells)[0]

            # print("Matched Buy with trade (ID: {}, DateTime: {})".format(matchingBuyById.ID, matchingBuyById.Date))
            # print("Matched Sell with trade (ID: {}, DateTime: {})".format(matchingSoldByDate.ID, matchingSoldByDate.Date))
        except StopIteration:
            print("Failed processing stock lot (ID: {}, ISIN: {}), found no match".format(input.ID, input.ISIN))
            raise StopIteration

        processed = pgf.TaxLotStock(
            ID=input.ID,
            ISIN=input.ISIN,
            Quantity=input.Quantity,
            Acquired=matchingBuyById,
            Sold=matchingSoldByDate,
            ShortLongType=cf.GenericShortLong.LONG,
        )

        return processed

File: lib/Core/StagingProcessor/Transformers/StockLotProcessor.ts
import { DateTime } from "luxon";
import type { ValidDateTime } from "@brrr/Utils/DateTime.ts";
import { FinancialIdentifier } from "@brrr/Core/Schemas/FinancialIdentifier.ts";
import { GenericShortLong } from "@brrr/Core/Schemas/CommonFormats.ts";
import { TradeEventStockAcquired, type TradeEventStockSold } from "@brrr/Core/Schemas/Events.ts";
import { TaxLot, type TaxLotStock } from "@brrr/Core/Schemas/Lots.ts";
import type { StagingTaxLot } from "@brrr/Core/Schemas/Staging/Lots.ts";
import { ProcessingUtils } from "@brrr/Core/FinancialEvents/ProcessingUtils.ts";

const utils = new ProcessingUtils();

export function processStockLot(
	input: StagingTaxLot,
	references: (TradeEventStockAcquired | TradeEventStockSold)[],
): TaxLotStock {
	const allBuys = references.filter((t): t is TradeEventStockAcquired => t instanceof TradeEventStockAcquired);
	const allSells = references.filter((t): t is TradeEventStockSold => !(t instanceof TradeEventStockAcquired));

	let matchingBuy: TradeEventStockAcquired;
	let matchingSell: TradeEventStockSold;

	try {
		matchingBuy = utils.findEventById(input.acquired.id ?? "", allBuys);
		const soldDate = input.sold.dateTime ?? DateTime.fromMillis(0) as ValidDateTime;
		const sellMatches = utils.findEventByDate(soldDate, allSells);
		if (sellMatches.length === 0) throw new Error("No sell found");
		matchingSell = sellMatches[0];
	} catch (_e) {
		console.error(`Failed processing stock lot (ID: ${input.id}, FinancialIdentifier: ${input.financialIdentifier}), found no match`);
		throw new Error(`LookupError: Failed to match stock lot (ID: ${input.id})`);
	}

	return new TaxLot({
		id: input.id,
		financialIdentifier: FinancialIdentifier.fromStagingIdentifier(input.financialIdentifier),
		quantity: input.quantity,
		acquired: matchingBuy,
		sold: matchingSell,
		shortLongType: GenericShortLong.LONG,
		provenance: [],
	});
}
